### PR TITLE
feat: implement web-tracing-framework support

### PR DESCRIPTION
**Pull request by mhevery**
This includes implementation and minimual

image

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2610)

### CODE REVIEW

File: modules/angular2/src/change_detection/abstract_change_detector.ts [hunk 1]
```diff
@@ -4,6 +4,9 @@ import {ChangeDetectorRef} from './change_detector_ref';
 import {ChangeDetector} from './interfaces';
 import {Locals} from './parser/locals';
 import {CHECK_ALWAYS, CHECK_ONCE, CHECKED, DETACHED, ON_PUSH} from './constants';
+import {enter2, leave, createScope} from 'angular2/src/core/wtf';
+
+var _scope_check = createScope(`ChangeDetector_#check(ascii id, bool throwOnChange)`);
```
Comment (vicb) [inline modules/angular2/src/change_detection/abstract_change_detector.ts]: `_scopeCheck` / `_checkScope` ?

File: modules/angular2/src/core/wtf.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ *          // do something
+ *        } finally {
+ *          leave(s);
+ *        }
+ *     }
+ *
+ * [executeInScope] helps with writing this boilerplate code.
+ */
+export function enter(scope: any): any {
+  return scope !== null ? scope(undefined, undefined) : null;
```
Comment (vicb) [inline modules/angular2/src/core/wtf.ts]: could it be `scope()` only ?

at least 1 & 2 `undefined` should probably be removed from `enter1` & `enter2` respectively

File: modules/angular2/src/core/zone/ng_zone.ts [hunk 3]
```diff
@@ -15,6 +17,9 @@ export interface NgZoneZone extends Zone { _innerZone: boolean; }
  * @exportedAs angular2/core
  */
 export class NgZone {
+  _zone_run = createScope(`NgZone#run()`);
```
Comment (vicb) [inline modules/angular2/src/core/zone/ng_zone.ts]: zone -> scope ?

File: modules/angular2/src/render/dom/dom_renderer.ts [hunk 4]
```diff
@@ -34,27 +36,40 @@ export class DomRenderer extends Renderer {
     this._document = document;
   }
 
+  _scope_createRootHostView = createScope('DomRenderer#createRootHostView()');
```
Comment (vicb) [inline modules/angular2/src/render/dom/dom_renderer.ts]: nit: move all properties before methods ?

File: modules/angular2/src/change_detection/abstract_change_detector.ts [hunk 5]
```diff
@@ -8,6 +8,10 @@ import {ProtoRecord} from './proto_record';
 import {Locals} from './parser/locals';
 import {Pipes} from './pipes/pipes';
 import {CHECK_ALWAYS, CHECK_ONCE, CHECKED, DETACHED, ON_PUSH} from './constants';
+import {wtfCreateScope, wtfLeave, ScopeEventFactory} from 'angular2/src/core/wtf';
```
Comment (mhevery) [inline modules/angular2/src/change_detection/abstract_change_detector.ts]: pull WTF into separate module

File: modules/angular2/src/change_detection/abstract_change_detector.ts [hunk 6]
```diff
@@ -8,6 +8,10 @@ import {ProtoRecord} from './proto_record';
...
+
+var _scope_check: ScopeEventFactory =
+    wtfCreateScope(`ChangeDetector_#check(ascii id, bool throwOnChange)`);
```
Comment (mhevery) [inline modules/angular2/src/change_detection/abstract_change_detector.ts]: remove `_`

File: modules/angular2/src/change_detection/abstract_change_detector.ts [hunk 7]
```diff
@@ -73,6 +79,7 @@ export class AbstractChangeDetector<T> implements ChangeDetector {
     this._detectChangesInShadowDomChildren(throwOnChange);
 
     if (this.mode === CHECK_ONCE) this.mode = CHECKED;
+    wtfLeave(s);
```
Comment (mhevery) [inline modules/angular2/src/change_detection/abstract_change_detector.ts]: add extra space

File: modules/angular2/src/core/wtf.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * negatively impact the performance of your application. For this reason we recommend that
+ * you don't add them to ensure that `wtfLeave` gets called. In production `wtfLeave` is a noop and
+ * so try-finally block has no value. When debugging perf issues, skipping `wtfLeave`, do to
+ * exception, will produce incorrect trace, but presence of exception signifies logic error which
+ * needs to be fixed before the app should be profiled. Add try-finally only when you expect that
+ * an exception is expected during normal execution while profiling.
+ *
+ */
+export var wtfCreateScope: (signature: string, flags?: any) => impl.ScopeEventFactory =
+    wtfEnabled ? impl.createScope : (signature: string, flags?: any) => (arg0?: any, arg1?: any) =>
```
Comment (mhevery) [inline modules/angular2/src/core/wtf.ts]: have the return value share instance to conserve memory.

File: modules/angular2/src/core/wtf.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Used to mark end of Scope.
+ *
+ * - `scope` to end.
+ * - `returnValue` (optional) to be passed to the WTF.
+ *
+ * Returns the `returnValue for easy chaining.
+ */
+export var wtfLeave:<T>(scope: any, returnValue?: T) => T = wtfEnabled ? impl.leave : (s, r?) => r;
```
Comment (mhevery) [inline modules/angular2/src/core/wtf.ts]: save instance

File: modules/angular2/src/core/wtf_impl.dart [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    case 1:
+        return ([arg0, arg1]) {
+          _arg1[0] = arg0;
+          return jsScope.apply(_arg1);
+        };
+    case 2:
+        return ([arg0, arg1]) {
+          _arg2[0] = arg0;
+          _arg2[1] = arg1;
+          return jsScope.apply(_arg1);
```
Comment (mhevery) [inline modules/angular2/src/core/wtf_impl.dart]: wrong array being passed here.

File: modules/angular2/src/core/wtf_impl.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  }
+  return false;
+}
+
+export function createScope(signature: string, flags: any = null): any {
+  return events.createScope(signature, flags);
+}
+
+export function leave<T>(scope: any, returnValue?: T): T {
```
Comment (mhevery) [inline modules/angular2/src/core/wtf_impl.ts]: add more types for TS

File: modules/angular2/src/render/dom/dom_renderer.ts [hunk 12]
```diff
@@ -50,19 +52,28 @@ export class DomRenderer extends Renderer {
...
+    var s = this._scope_createRootHostView();
     var hostProtoView = resolveInternalDomProtoView(hostProtoViewRef);
     var element = DOM.querySelector(this._document, hostElementSelector);
     if (isBlank(element)) {
+      wtfLeave(s);
       throw new BaseException(`The selector "${hostElementSelector}" did not match any elements`);
     }
-    return this._createView(hostProtoView, element);
+    var view = this._createView(hostProtoView, element);
+    wtfLeave(s);
```
Comment (mhevery) [inline modules/angular2/src/render/dom/dom_renderer.ts]: rewrite as `return wtfLeave(s, view)`;

File: modules/angular2/src/render/dom/dom_renderer.ts [hunk 13]
```diff
@@ -50,19 +52,28 @@ export class DomRenderer extends Renderer {
...
+    return view;
   }
 
+  _scope_createView = wtfCreateScope('DomRenderer#createView()');
   createView(protoViewRef: RenderProtoViewRef, fragmentCount: number): RenderViewWithFragments {
+    var s = this._scope_createView();
     var protoView = resolveInternalDomProtoView(protoViewRef);
-    return this._createView(protoView, null);
+    var view = this._createView(protoView, null);
+    wtfLeave(s);
```
Comment (mhevery) [inline modules/angular2/src/render/dom/dom_renderer.ts]: same here

**Pull request closed without merging**